# Gene expression metaprograms of cancer cells

**Related figures**: Fig. 3, Fig. S3. 


**Rscript files**: 

- <kbd>analysis/scripts/xxx.R</kbd> ([link](https://github.com/navinlabcode/tnbc-chemo/blob/main/analysis/scripts/xxx.R)). 

## Identifying metaprograms of cancer cells

## Determining cellular frequencies of metaprograms

**Synopsis**

``` console
Rscript 
```

| Parameter             | Meaning                                                 |
| --------------------- | ------------------------------------------------------- |



**Output**

**Rationale**

*Why not computationally integrate all cells?*

*Why not perform NMF on all single cells?*

*Why not use consensus NMF ([Kotliar, et al. 2019](https://doi.org/10.7554/eLife.43803))?*

>[!NOTE]
> *A possible limitation of metaprogram analysis*
>
>
>
>
>
>







